daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueCoder.scala

```diff
@@ -46,61 +46,139 @@ object ValueCoder {
-          AbsoluteContractId
-        )
-    override def fromStruct(
-        s: String,
-        isRelative: Boolean): Either[DecodeError, AbsoluteContractId] =
-      if (isRelative)
-        Left(DecodeError(s"unexpected relative contractId $s"))
-      else
-        fromString(s)
+  abstract class DecodeCid[Cid] private[lf] {
+    def decodeOptional(
+        sv: SpecifiedVersion,
+        stringForm: String,
+        structForm: proto.ContractId,
+    ): Either[DecodeError, Option[Cid]]
+
+    final def decode(
+        sv: SpecifiedVersion,
+        stringForm: String,
+        structForm: proto.ContractId,
+    ): Either[DecodeError, Cid] =
+      decodeOptional(sv, stringForm, structForm).flatMap {
+        case Some(cid) => Right(cid)
+        case None => Left(DecodeError("Missing required field contract_id"))
+      }
   }
 
-  object CidDecoder extends DecodeCid[ContractId] {
+  val CidDecoder: DecodeCid[ContractId] = new DecodeCid[ContractId] {
 
-    private def fromStringToRelCid(s: String): Either[DecodeError, RelativeContractId] =
+    private def stringToRelativeCid(s: String) =
       scalaz.std.string
         .parseInt(s)
         .toEither
         .bimap(
-          _ => DecodeError(s"cannot parse relative contractId $s"),
-          idx => RelativeContractId(NodeId(idx), None)
+          _ => //
+            DecodeError(s"""cannot parse relative contractId "$s""""),
+          idx => Some(RelativeContractId(NodeId(idx), None))
         )
 
-    override def fromString(s: String): Either[DecodeError, ContractId] =
-      if (s.startsWith("~")) fromStringToRelCid(s.drop(1)) else AbsCidDecoder.fromString(s)
+    private def stringToCidString(s: String): Either[DecodeError, Ref.ContractIdString] =
+      Ref.ContractIdString
+        .fromString(s)
+        .left
+        .map(_ => //
+          DecodeError(s"""cannot parse absolute contractId "$s""""))
+
+    override def decodeOptional(
+        sv: SpecifiedVersion,
+        stringForm: String,
+        structForm: proto.ContractId,
+    ): Either[DecodeError, Option[ContractId]] =
+      if (useOldStringField(sv)) {
+        if (structForm != proto.ContractId.getDefaultInstance) {
+          Left(DecodeError(s"${sv.showsVersion} is too new to use string contract IDs"))
+        } else {
+          if (stringForm.isEmpty)
+            Right(None)
+          else if (stringForm.startsWith("~"))
+            stringToRelativeCid(stringForm.drop(1))
+          else
+            for {
+              coid <- stringToCidString(stringForm)
+              _ <- Either.cond(
+                test = Ref.ContractIdString.isA(coid),
+                right = (),
+                left = DecodeError(sv, s"absolute contractId V1"),
+              )
+            } yield Some(AbsoluteContractId(coid))
+        }
+      } else {
+        if (stringForm.nonEmpty)
+          Left(DecodeError(sv, isTooOldFor = "message ContractId"))
```

Comment:
The error message should say that string form shouldn't be used.

Reply:
You are right